%\documentclass[12pt]{article}
\documentclass[a4paper,10pt]{amsart}
\usepackage{ucs}
\usepackage[utf8x]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{graphicx}
\usepackage{sidecap}
%\setlength{\captionindent}{0pt}
\usepackage{url}
\usepackage{amsmath}
\usepackage{Sweave}
%\usepackage{natbib}
\usepackage{graphicx}

% similar to Matrix package
\usepackage[authoryear,round]{natbib}

%\bibliographystyle{plainnat}
%\bibliographystyle{harvard}
\bibliographystyle{mychicago}

\SweaveOpts{engine=R,eps=FALSE,pdf=TRUE,width=5,height=3,strip.white=true,keep.source=TRUE}
<<preliminaries, echo=FALSE>>=
options(digits=4, width=60)
@


%\textwidth=16cm
%\textheight=24cm
%\parskip=0.3cm
%\topsep=0cm
%\oddsidemargin=0cm
%\topmargin=-0.3cm
%\headsep=0cm
%\headheight=0cm
%\columnsep=0.5cm

\def\ccDesign{\texttt{ccDesign}}
\def\ldDesign{\texttt{ldDesign}}
\def\BayesQTLBIC{\texttt{BayesQTLBIC}}
\def\Pr{\mbox{Pr}}
\def\case{\mbox{case}}
\def\control{\mbox{control}}
\def\data{\mbox{data}}
\def\bb{\texttt{bb}}
\def\Bb{\texttt{Bb}}
\def\BB{\texttt{BB}}
\def\ie{\textit{i.e.}}
\def\eg{\textit{e.g.}}
\def\etal{\textit{et al.}}
\def\cf{\textit{cf.}}

\setcounter{secnumdepth}{4}
\setcounter{tocdepth}{4}
\begin{document}

%\VignetteIndexEntry{Using lDesign}

\title{\ldDesign{} 2: Experimental design for genome-wide association studies}
\author{
\bf Roderick D. Ball
}
\date{}
\maketitle
\vspace*{5.5in}
Copyright \copyright 2011  by Roderick D. Ball, Scion/VISG
\pagebreak
\tableofcontents

\section{Introduction}

This vignette illustrates the use of the \ldDesign{} R package for
design of experiments for detecting genomic associations.  Version 1
of this package implements the method for quantitative traits (Ball
2005: ``Experimental designs for reliable detection of linkage
disequilibrium in unstructured random population association studies''
Genetics 2005). There are two main functions: \texttt{ld.power()} and
\texttt{ld.design()} for determining the power and sample size
respectively.  Version 2 of this package also implements the method
for case-control studies (Ball 2011; ``Experimental designs for robust
detection of effects in genome-wide case-control studies'' Genetics
2011).  For case-control studies there are two main functions:
\texttt{cc.power()} and \texttt{cc.design()} for determining the power
and sample size respectively.  

While \ldDesign{} was written primarily for genetic applications,
\ldDesign{} also provides generally useful functions like
\texttt{SS.oneway.bf()} for calculating Bayes factors from one-way
analysis of variance models, \texttt{oneway.bf.alpha()} for
calculating the Bayes factor for a given $\alpha$ level in one-way
analysis of variance models, \texttt{calc.Balpha.ABF()} for
calculating the approximate Bayes factor corresponding to a given
$\alpha$ level and \texttt{calc.alphaB.ABF()} for calculating the
$\alpha$ level corresponding to an approximate asymptotic Bayes
factor.  These functions are useful for retrospective calculations:
\eg{} determining the Bayes factor when $p$-values are published, or to
give Bayesian measures of evidence when using frequentist methods like
analysis of variance or \textit{vice versa}.  Readers interested in these
functions for non-genetic analyses may wish to skip to
Section. \ref{sec:bayesianfunctions}.

This package implements a Bayesian experimental design methodology
designed to ensure robust detection of effects in genome-wide
association studies (GWAS).  In GWAS, hundreds of thousands of loci
are being tested on thousands of individuals. To robustly detect
effects we need sufficient power to detect the effects with
sufficiently strong evidence to overcome the low prior odds for any
given polymorphism to be associated with a detectable size effect.

Traditional approaches design experiments with power to detect effects
with a given $p$-value or less ($P < \alpha$ for some $\alpha$).
However the $p$-value as a measure of evidence can be misleading and
is not well calibrated for detecting effects in general. This is
particularly problematic in genome-wide association studies where
sample sizes are large and effect sizes are small. Traditional
approaches using $p$-values have lead to many published spurious
associations
\citep{Altshuleretal2000,TerwilligerWeiss1998,Emahazionetal2001,
  Ball2005exptdesignbf,Ball2007AssocMapInPlantsCh8,Ball2007QTLcolocation,
  Ball2011exptdesigngwcc}
%(Altshuler {\it et al.} 2000; Terwilliger and Weiss 1998; Emahazion {\it et al.} 2001; Ball 2005, 2007a, 2007b, 2011).
Even very low $p$-values (\eg{} $5\times 10^{-7}$,
or $5\times 10^{-8}$) may not be sufficient for respectable posterior
odds in large studies or meta-analyses \citep{Ball2011exptdesigngwcc}.

Traditional frequentist approaches adjust for multiple tests.
However, gene mapping is, intrinsically, a model selection problem not
a hypothesis testing problem. The model we wish to determine is
ultimately something approximating the true genetic model---the set of
loci affecting the trait and their modes of action.  See \eg{} 
\citep[and references therein]{Ball2001,SillanpaaCorander2002}

%(Ball 2001; Sillanp\"a\"a and Corander 2002 and references therein). 

Often the true model is not unequivocally determined by the data.
Neglecting this leads to problems with spurious associations and
selection bias, where the estimated effects are over-estimated unless
the power to detect the true size of effect is high or independent
data is used for detection and estimation. We recommend Bayesian
`model selection' approaches that consider alternative possible models
according to their probabilities (\eg{} \cite{Ball2001}). A
non-MCMC implementation of Bayesian model selection for QTL mapping is
given in the R package \BayesQTLBIC{} \citep{Ball2009BayesQTLBIC}.

The power calculations in this package are for single marker tests,
which are widely used. This will be a good approximation as long as
the resolution of the study and the extent of linkage disequilibrium
is such that effects are approximately independent. When multiple
markers within the extent of linkage disequilibrium are affecting the
trait multi-locus methods for analysis are recommended. 
%(\eg{} Ball 2001; R package \BayesQTLBIC{}). 
In this case, the power calculations in \ldDesign{} would
be approximate and conservative.  

This package uses a novel approach: unlike traditional power
calculations that use $p$-values as a measure of evidence, we use
Bayes factors, which, unlike $p$-values, have a direct interpretation
as strength of evidence independent of sample size, experimental
design, and test set-up. Designing experiments with power to obtain a
sufficiently large Bayes factor, $B$, where $B$ is chosen large enough
to obtain respectable posterior odds enables us to design experiments
with sufficient power to robustly detect associations at unknown
positions in the genome. 

Note that by obtaining the Bayes factor in single marker tests we are still
considering multiple models: corresponding to the null and alternative
hypotheses $H_0, H_1$ --- the Bayes factor combined with prior odds
determines the posterior probabilities. If posterior odds are not high
an unbiased estimator would be the product of the estimate conditional
on $H_1$ and the posterior probability for $H_1$. This would often result
in considerable shrinkage of effect estimates.

Association studies detect \textit{linkage disequilibrium} between an
observed marker locus and an unobserved trait locus (QTL or QTN).
Since linkage disequilibrium between 2 loci decays exponentially each
generation at a rate proportional to the recombination rate between
the loci, association mapping using population based samples offers
potentially higher resolution (resolution 100s to 1000s of base
pairs) than family based QTL mapping approaches (centi-Morgans to 10s
of centi-Morgans) which detect linkage disequilibrium generated within
a pedigree.  However achieving this resolution requires much larger
sample sizes.  Spurious associations can be generated by population
structure or simply because the strength of evidence is insufficient
to overcome the low prior odds for genomic associations in diverse
populations.  For further information see 
\citep[and references therein]{Ball2005exptdesignbf,Ball2007AssocMapInPlantsCh8,Ball2007QTLcolocation}

%(Ball 2005, 2007a; 2011, and references therein).


\section{Quantitative traits}
\subsection{Factors affecting power}

For a given marker and causal locus, factors affecting power of an experiment
to detect linkage disequilibrium between the marker and causal locus include:
\begin{itemize}
\item sample size ($n$)
\item allele frequency at the marker locus ($p$)
\item allele frequency at the causal locus ($q$)
\item linkage disequilibrium coefficient ($D$, or $D'$)
\item effect QTL heritability ($h^2_q$)
\item genetic model (dominance ratio, $\phi$)
\item Bayes factor required ($B$)
\end{itemize}
corresponding to the arguments of \texttt{ld.power()}:

<<>>=
library(ldDesign) 
args(ld.power) 
@

The function \texttt{ld.power()} implements the method from
\cite{Ball2005exptdesignbf} which uses the \cite{SpiegelhalterSmith82}
Bayes factor---an analytical formula for one way analysis of variance
models (implemented in the R function \texttt{SS.oneway.bf()}, \cf{}
Section~\ref{sec:bayesianfunctions}), in conjunction with a frequentist
power calculation adapted from \cite{Luo1998}.

Note: the \cite{SpiegelhalterSmith82} Bayes factor---an analytical
formula for one way analysis of variance models (function
\texttt{SS.oneway.bf()}), so does not incorporate prior variance. In
our experience, use of \texttt{SS.oneway.bf()} is approximately
equivalent to assuming prior information equivalent to a single sample
point ($a=1$ in the arguments to \texttt{cc.power()} below). 
In principle \texttt{ld.power()} could incorporate prior variance using 
the methods used by \texttt{cc.power()}. This may be incorporated in
a future version of \texttt{ld.power()}.

\subsection{Bayes factor, $B$}

The Bayes factor for comparing two models ($M_0,M_1$) is the ratio 
\begin{gather}
B = \frac{\Pr(\data\mid M_1)}{\Pr(\data\mid M_0)}
\end{gather}

Prior and posterior odds are related by:
\begin{gather}
\mbox{posterior-odds} = B \times \mbox{prior-odds}
\end{gather}

Hence, \eg{} if the prior odds were 1:50000 and we want posterior odds of
20:1, we should have a Bayes factor of
\begin{gather}
\frac{20}{1/50,000} = 1,000,000
\end{gather}

\subsection{Linkage disequilibrium}

We do not necessarily observe the causal locus, but a marker-trait
association is induced by \textit{linkage disequilibrium} between the
marker and a causal locus.
Linkage disequilibrium  \citep{Weir1996GeneticDataAnalysisII}
is non-independence between genetic
loci or positions on the genome. We will consider bi-allelic loci,
\eg{} a marker with alleles $A,a$ and a causal locus with alleles $Q,q$.
The pairwise probabilities for the alleles
\begin{align}
\Pr(AQ) &= \Pr(A)\Pr(Q) + D &= pq + D  \\
\Pr(aQ) &= \Pr(a)\Pr(Q) - D &= (1-p)q - D \\
\Pr(Aq) &= \Pr(A)\Pr(q) - D &= p(1-q) - D \\
\Pr(aq) &= \Pr(a)\Pr(q) + D &= (1-p)(1-q) + D
\end{align}
where $D$ is the linkage disequilibrium coefficient and $p,q$ are the
allele frequencies at the marker and causal locus,
respectively. Linkage disequilibrium can also be specified as $D'$
which is $D$ expressed as a proportion of the maximum (resp. minimum)
disequilibrium if $D$ is positive (resp. negative).

It is often more convenient to specify $D'$ ($D$ divided by its
maximum absolute value for the given allele frequencies and sign of
$D$) because otherwise we have to work out the maximum or minimum
values of $D$ for the given allele frequencies.

Another useful quantity is $r^2$, related to $D,p,q$ by:
\begin{equation}
\label{eq:rsquared}
r^2 = \frac{D^2}{p(1-p)q(1-q)}
\end{equation}
Similar to $r^2$ in linear regression, $r^2$ gives approximately the
proportion of variance explained by the marker in predicting the
causal locus.  The approximation is to first order for $r^2 \approx
1$, however in practice $r^2$ may be significantly less than 1. For
this reason our power calculations use exact quantities in expressed in terms of
$D,p,q$.

\subsection{Examples}
\begin{itemize}
\item[1.] Find the power to detect an effect with QTL heritability
  $h^2_q = 0.05$, with   Bayes factor $10^6$, marker and causal allele
  frequencies 0.3, 0.2,  linkage disequilibrium coefficient $D=0.1$, 
  and sample size 1000.  Assume an additive model.

<<>>=
ld.power(B=1e6, h2=0.05, D=0.1, p=0.3, q=0.2, n=1000, phi=0)
@
\item[2.] Find the power to detect an effect with QTL heritability
  $h^2_q = 0.05$, with   Bayes factor $10^6$, marker and causal allele
  frequencies 0.3, 0.2,  linkage disequilibrium coefficient $D=0.1$, 
  and sample size 1000.  Assume a dominant model.

<<>>=
ld.power(B=1e6, h2=0.05, D=0.1, p=0.3, q=0.2, n=1000, phi=1)
@
\item[3.] As per [2.] above find the sample size required for power 0.8 and print the
power curve:
<<>>=
ld.design(B=1e6, h2=0.05, D=0.1, p=0.3, q=0.2, phi=1, power=0.8, 
          nmin=1730, nmax=4620, ninterp=20, print.it=TRUE)
@
\end{itemize}

\section{Case-control studies}

\subsection{Factors affecting power}

For a given marker and causal locus, factors affecting power of an experiment
to detect linkage disequilibrium between the marker and causal locus include:
\begin{itemize}
\item Bayes factor required ($B$)
\item odd ratio(s) (OR) (or relative risk(s) ($R$))
\item linkage disequilibrium coefficient ($D$, or $D'$)
\item allele frequency at the marker locus ($p$)
\item allele frequency at the causal locus ($q$)
\item disease prevalence (or baseline risk)
\item sample sizes (number of cases, number of controls)
\item genetic model (additive, dominant, recessive or general)
\item prior variance for effects ($a$, or $\sigma^2_\eta$)
\end{itemize}
corresponding to the arguments of \texttt{cc.power()}:
<<>>=
library(ldDesign)
args(cc.power)
@
%\begin{verbatim}
%cc.power(B, OR, D, p, q, baseline.risk, Dprime=NULL, R=NULL, prevalence=NULL, 
%	 n.cases, n.controls, model=c("additive","dominant","recessive","general"), 
%	 a=1, sigma2.eta=NULL, verbose=FALSE, amalgamate.cells=FALSE)
%\end{verbatim}

\subsection{Odds ratios}
The odds ratio for 2 genotypes $g_1, g_2$ is the ratio
\begin{gather}
\frac{\Pr(\case\mid g_2)/\Pr(\control\mid g_2)}{\Pr(\case\mid g_1)/\Pr(\control\mid g_1)}
\end{gather}

For the additive, dominant and recessive models a single odds ratio or relative risk is
specified. For the general model a vector of 2 odds ratios or relative risks are specified.

\subsection{Examples}
\begin{itemize}
\item[1.] Find the power to detect an effect with odds ratio 1.6, with
  Bayes factor $10^6$, marker and causal allele frequencies 0.3, 0.2,
  linkage disequilibrium coefficient $D=0.1$, baseline risk 0.1, 1000
  cases and 1000 controls.  Assume an additive model, and a prior with
  information equivalent to a single sample point.

<<>>=
cc.power(B=1e6, OR=1.6, D=0.1, p=0.3, q=0.2, baseline.risk=0.1, 
         n.cases=1000, n.controls=1000, model="additive", a=1)
@

\item[2.] (Illustrating vectorisation of sample sizes).
As per [1.] above but find the power for sample sizes
from 2000 to 12000 in steps of 2000.

<<>>=
cc.power(B=1e6, OR=1.6, D=0.1, p=0.3, q=0.2, baseline.risk=0.1, 
         n.cases=1000*seq(2,12,by=2), 
         n.controls=1000*seq(2,12,by=2),
         model="additive", a=1)
@
\item[3.] Find the sample size required
to detect an effect with odds ratio 2.0 with Bayes factor $10^6$,
marker and causal allele frequencies 0.3, 0.2,
linkage disequilibrium coefficient $D=0.1$, baseline risk 0.1.
Assume the number of controls is 50\% more than the number of cases.
Assume an additive model, and a prior with information equivalent to
a single sample point.
Print the power curve with power ranging from 0.1 to 0.99.

<<>>=
cc.design(B=1e6, OR=2.0, D=0.1, p=0.3, q=0.2, power=0.9, 
          baseline.risk=0.1, n.cases=2000, n.controls=3000, 
          model="additive", a=1, pmin=0.1, pmax=0.99, 
          ninterp=20, print.power.curve=TRUE)
@
\item[4.] As per [3.] but assume a general model with 2 independent odds ratios
of 1.5, 1.5.
<<>>=
cc.design(B=1e6, OR=c(1.5,1.5), D=0.1, p=0.3, q=0.2, power=0.9, 
          baseline.risk=0.1, n.cases=2000, n.controls=3000, 
          model="general", a=1, print.power.curve=FALSE)
@
\item[5.] (Show attributes.) As per [1.] show attributes
including non-centrality parameter (\texttt{ncp}) and the optimal
weighing used (\texttt{c1.opt}).
<<>>=
cc.power(B=1e6, OR=1.6, D=0.1, p=0.3, q=0.2, baseline.risk=0.1, 
         n.cases=1000, n.controls=1000, model="additive", a=1,
         show.attributes=TRUE)
@

A brief explanation of the attributes is shown in Table~\ref{tbl:ccpowerattr}. 
Other attributes are as given in the function call, except that odds ratios and
relative risks are both shown and baseline risk and prevalence are both shown regardless
of which were specified.
\end{itemize}

\begin{table}[htbp]
\caption{Attributes returned by \texttt{cc.power()}}
\label{tbl:ccpowerattr}
\begin{tabular}{|l|p{6cm}|}
\hline\hline
\texttt{ncp} &  the non-centrality parameter (12.63) \\
\hline
\texttt{alphac} &  the value of $\alpha$ threshold corresponding to the
given Bayes factor ($7.73\times 10^{-9}$) \\
\hline
$\texttt{c1.opt}$ & $\texttt{c1.opt} = 0.679$  is the weight placed on the first odds ratio
(\Bb{} vs \bb)
and $1 - \texttt{c1.opt}$ = 0.321 is the weight placed on the second odds ratio
(\BB{} vs \Bb) when estimating the odds ratio in the additive model\\
\hline
\texttt{ps} &  table of expected marker allele frequencies. A 2x3 table
for the additive or general models and a 2x2 table (with functionally equivalent
genotypes amalgamated) for the dominant or recessive models. \\
\hline
\end{tabular}
\end{table}

\section{Design of Association Studies}

Here we discuss application to design of GWAS for detecting genomic
associations, including elicitation of the prior parameters.

\subsection{Genomic associations}

The genome contains many (\eg{} $3\times 10^9$ for humans) loci.
Typically about 1/1000 loci are polymorphic, \ie{} differ between individuals
in a species. For a given trait, only a small proportion or polymorphic loci 
will be causal loci with practically significant effects on the trait.

\subsection{Genome-wide association studies}

Genome-wide association studies (GWAS) aim to detect associations
between SNP markers spaced along the genome and causal loci. If a
marker is in linkage disequilibrium with a causal locus, then the
marker genotypes will be associated with trait variation. The strength
of association depends on the size of the causal effect and the
linkage disequilibrium coefficient $D$.

\subsection{Prior elicitation}

Prior odds and prior variance for the effect(s) being tested are
critical factors in determining the Bayes factor
required.  Ascertaining the prior is, inevitably, subjective. This
does not mean, however, that we can `choose' the prior
arbitrarily. The prior should represent our prior knowledge before
carrying out the experiment or observing the data. This is a subjective
Bayesian approach as opposed to `objective' Bayesian approaches that 
try to use a `default' or non-informative prior. Using $a=1$ in
\texttt{cc.power()} is an example of a default prior. While it is useful
to have a good conservative default it is not recommended in general to always
use this. The process of ascertaining the prior is known as prior elicitation.

Prior elicitation \citep{OHagan2006prior_elicitation_book},
%(O'Hagan \textit{et al.} 2006)
where prior information is elicited from experts is an important but neglected
area. This is likely to become more important (and interesting) as
more associations are detected.

In general the prior odds depend on the number of loci expected to
significantly affect the trait (ignoring effects that are
undetectable), the number of markers and the extent of linkage
disequilibrium.  If marker spacing is comparable to the extent of LD,
then we consider the prior probability per interval of this size
around a marker.  If there are fewer markers the number of markers is
limiting, and we again consider the prior probability per interval of
this size around a marker.  If there are substantially more markers we
may increase the minimum $D$ that we design for.

Prior odds also depend on previous results from the same or similar
loci and/or traits in the same or related populations and/or species.
In practice one never observes true replicates but experiments, data
or information that have varying degrees of relevance to the
experiment or decision in question. What weight to give various
elements of information is the subjective aspect, where experts can be
useful. This is central to the application of statistics in general
since we are always trying to infer or predict something in a new or
similar but different situation using information from past
observations.

\subsection{Genome scans}
\label{sec:genomescans}
First consider the scenario as in Table~\ref{tbl:scenario1} with approximately
one marker per interval of length equal to the extent of LD.

\begin{table}[htbp]
\caption{Determination of prior odds and Bayes factor required.}
\label{tbl:scenario1}
\begin{tabular}{|l|r|}
\hline\hline
Bayes factor from previous studies & 500 \\
genome length           &  $3\times 10^9$ \\
expected number of loci & 10 \\
extent of LD            & 6kb \\
number of SNP markers   & 500000 \\
prior odds per marker   & $6\times 10^3/3\times 10^9 = 1/500000 $ \\
posterior odds required &  20:1\\
combined Bayes factor required   &  $10^6$\\
study Bayes factor required & $10^6/500 = 2000$\\
\hline
\end{tabular}
\end{table}

\defcitealias{HapMap}{\url{www.hapmap.org}}
\defcitealias{WTCCC2007}{WTCCC2007}

Note:
\begin{itemize}
\item The posterior probability per marker is the probability that the marker
is within the extent of LD of a causal marker.

\item We are free to choose the criterion for defining the extent of LD,
\eg{} $D'> 0.5$ or $r^2 > 0.8$. However if choosing a lower threshold
this will reduce power, and if choosing a higher threshold this will
result in smaller sized intervals requiring more markers to be
genotyped for acceptable coverage.
\item Conversely if we have more markers we may use a higher value of $D$ 
in the power calculations.
\item The expected number of loci, should be the expected number of
  loci with detectable size effects. If the experiment is sufficiently
  powerful the expected number of effects may increase. For example, if the
  intended experiment is powerful enough to detect effects explaining 1\% of
  the phenotypic variance and the trait heritability is 30\%, there
  can be at most 30 such effects. In practice the amount of variation
  explained by additive effects of 1\% or greater may be less,
  \eg{} only 10\%, due to the possible existence of many smaller
  effects, and non-additive variation, justifying and expected number of loci
  of 10.
\item We also need to consider the allele frequencies for loci we wish
  to detect. Rare alleles may be poorly tagged by common SNPs
  \eg{} from the HapMap project \citepalias{HapMap} (e.g
  \cite{YangetalcommonSNPs2010}).  Large recessive effects may not yet
  be detected \cite{Ball2011exptdesigngwcc}, \eg{}  to detect an effect
  of odds ratio 3 with minor allele frequency 5\% when tagged by a
  marker with allele frequency 10\% at $D'=0.5$ requires over $10^6$
  cases and controls: 
<<>>= 
cc.design(B=1e6, OR=3, Dprime=0.5, p=0.1, q=0.05, 
          baseline.risk=0.1, n.cases=1000, n.controls=1000,
          model="recessive", a=1, power=0.8, 
          print.power.curve=FALSE) 
@
\end{itemize}

Recent analyses for human height (\eg{} \citet{Gudbjartssonetal2008},
\citet{LangoAllen2010},\citet{Weedonetal2008}) have of the order of
15,000 samples in combined meta-analyses. Hundreds of putative effects
were identified although few were replicated across all three analyses.
For example, \cite[Table 1, p577]{Weedonetal2008} reported 20 SNP
effects with combined $p$-values $P < 5\times 10^{-7}$ putatively
collectively explaining 3\% of the variance from a meta-analysis of
human height. We re-examine the power, using a prior odds based on 100 expected loci:

<<>>=
# power for a well tagged locus
ld.power(B=5e5, h2=0.001, D=0.25, p=0.5, q=0.5, n=13665, phi=1) 
@
%# power for a less well tagged locus with D=Dmax, MAF 0.05
%ld.power(B=5e5,h2=0.001, D=0.045, p=0.1, q=0.05, n=13665, phi=1) 
%# power for a less well tagged locus with D=Dmax/2, MAF 0.05
%ld.power(B=5e5,h2=0.001, D=0.045/2, p=0.1, q=0.05, n=13665, phi=1) 
%@
we see the power is low to obtain respectable posterior odds, even for
a well tagged locus, hence these effects are not robustly detected. 
At the very least, the effects need to be re-estimated in an independent population.

The  Bayes factor corresponding to the threshold used is:
<<>>=
calc.Balpha.ABF(alpha=5e-7, n=13665, a=1)
@
representing strong evidence, but not strong enough for respectable
posterior odds without replication at a similar strength of evidence.

The sample size required for good power is from 50,000 for a well
tagged locus ($r^2 \approx 1$) or 110,000 for a less well tagged locus
($r^2=0.5$) or 460,000 for a poorly tagged locus ($r^2=0.12$):
<<>>=
# sample size for a well tagged locus
ld.design(B=1e5, h2=0.001, D=0.25, p=0.5, q=0.5, power=0.8, 
          phi=1) 
# sample size for a less well tagged locus, D=Dmax, MAF=0.05
ld.design(B=1e5, h2=0.001, D=0.045, p=0.1, q=0.05, power=0.8, 
          phi=1, nmin=50000, nmax=200000)
# sample size for a poorly tagged locus, D=Dmax/2, MAF=0.05
ld.design(B=1e5, h2=0.001, D=0.045/2, p=0.1, q=0.05, power=0.8, 
          phi=1, nmin=50000, nmax=500000)
@

Note the use of arguments \texttt{nmin} and \texttt{nmax} here. These values should
be chosen to bracket the required sample size, as interpolation is used on values
calculated by \texttt{ld.power()} for sample sizes from \texttt{nmin} to \texttt{nmax}.
This may require some iteration. Having found a solution, the solution can be refined by
specifying \texttt{nmin} and \texttt{nmax} closer to the solution.


%and are reporting putative effects explaining 0.1 0.2\% (check?) of the variance.
%In this case we could allow for a prior expected number of effects of 100 explaining
%more than 0.1\% of the variation in
%Table~\ref{tbl:scenario1}, meaning we need a Bayes factor of $10^5$. The power
%for a well tagged locus would be 
%<<>>=
%ld.power(B=1e5,h2=0.001, D=0.25, p=0.5, q=0.5, n=100000, phi=1)
%@
%The power for a less well tagged locus with low allele frequency at half the maximum
%LD ($D=0.045, p=0.1, q=0.05, r^2=0.47$) would be
%<<>>=
%ld.power(B=1e5,h2=0.001, D=0.045, p=0.1, q=0.05, n=100000, phi=1)
%@
%by comparison the if there were 10 loci each explaining 1\% of the variance the power 
%would be
%<<>>=
%ld.power(B=1e6, h2=0.01, D=0.25, p=0.5, q=0.5, n=100000, phi=1)
%ld.power(B=1e6, h2=0.01, D=0.045, p=0.1, q=0.05, n=100000, phi=1)
%@



\subsection{Candidate gene studies}

\ldDesign{} applies equally well to candidate gene studies, where
polymorphisms are sourced from `candidate genes' considered likely to
affect the trait. Depending on the trait and candidate gene, the
polymorphism may have higher prior probability than a random genomic
polymorphism.

As a default starting value we may assume similar prior odds to a
genome scan.

Substantially higher prior odds than this require some
justification. Just being in a gene or associated promoter alone does
not necessarily increase prior odds by much since there are $\sim
50,000$ genes and a number (\eg{} up to $\sim 10$ ??) of polymorphisms
in each gene, and we have to allow for the possibility that a
proportion of causal effects are not in the candidate gene set.

A polymorphism mapping into a QTL region may have higher prior odds
depending on strength of evidence for the QTL, the posterior
distribution for QTL location, and relative position of the
polymorphism and the QTL \citep{Ball2007QTLcolocation}.

Bioinformatics, \eg{} where similar sequences have been found
in other species, and knowledge of gene action and pathways that 
affect the trait, may be useful.

\subsection{Dense markers and whole genome sequencing}

With dense markers or whole genome sequencing we may be in the
luxurious position of having multiple or many markers within the
extent of LD of any given marker or causal locus. This means there
may be multiple markers within the extent of LD of a causal locus
competing to explain the variation. A work-around is to consider the
prior odds for a single representative marker within the interval.

Whole genome sequencing raises additional challenges \eg{} errors in
marker genotyping when coverage is low or limited sample sizes when
coverage is high. The argument \texttt{missclass.rate} in
\texttt{ld.power()} can be used where there is an estimate of
genotyping error rate.

On the `$+$' side of the ledger having whole genome sequences means we
will have the causal locus, \ie{} some polymorphism in complete LD with
a causal locus.  We just have to identify which ones. This should give
more power, quantifying this is a topic for future research.

\subsection{Choice of population}

Where sample size is limited experimenters may be better off choosing
a less diverse population, \eg{} an isolated Finnish or island
population with a relatively small number of founders, rather than a
diverse African population.

The tradeoff is lower resolution but smaller sample size and number of
markers needed to obtain that resolution. This means more power to
detect effects but a number of effects (particularly rare variants)
may not be present in the small population. If studying a disease, the
disease would obviously have to be present in the study population. A
population with high prevalence may be promising.

Effects could then be validated in other populations requiring genotyping only
a limited number of markers.

\subsection{Choice of trait} 

A trait closer to basic biochemistry \eg{} resulting from a known
biochemical pathway is likely to be influenced by only a moderate
number of 
%(mostly known?) 
genes. Prior odds would be much higher for
polymorphisms in the known genes.

Human height or traits such as tree growth rate or wood density may be
influenced by many small effect genes. Putative associations from GWAS
to date explain only a small proportion of the variation in human
height and complex diseases.

Some of these may in fact be larger effect rare alleles poorly tagged
by current SNPs (\cf{}  the example in subsection
\ref{sec:genomescans}).


\subsection{Multi-stage trials and replication}

\ldDesign{} does not yet cater explicitly for multi-stage
designs. However, when detecting and validating effects in several
stages we may partition the Bayes factor required among stages,
\eg{} if $B > 1000$ in each of 2 independent samples this combines,
conservatively, to give $B > 10^6$, of the order required for genomic
associations.


It is not sufficient just to say the results have been replicated for
some $\alpha$ level (which is usually much less than the $5\times 10^{-7}$ used
in genome scans circa 2007 (\eg{} \citetalias{WTCCC2007})).  It is vital that
the replication sample(s) have a sufficiently large Bayes factor so that
the combined Bayes factor is large enough to achieve the required
posterior odds. Since the $p$-value $5\times 10^{-7}$ corresponds to
a Bayes factor around $1000$ this value replicated \textit{twice} would be of the
order of magnitude needed. 

The current \textit{de facto} standard of $5\times 10^{-8}$ is
the threshold for the combined $p$-value over all replicates. This
 corresponds to a Bayes factor of around 24000 for \cite{Weedonetal2008}($n=13665$) or
a Bayes factor of around 9000 if $n=100,000$:
<<>>=
calc.Balpha.ABF(alpha=5e-8, n=13665, a=1)
calc.Balpha.ABF(alpha=5e-8, n=100000, a=1)
@



\section{Functions useful for general Bayesian analysis}
\label{sec:bayesianfunctions}  

\subsection{\texttt{SS.oneway.bf()}} 
This function, used by \texttt{ld.power()}, calculates the
\cite{SpiegelhalterSmith82} Bayes factor corresponding to a one-way
analysis of variance with given group sizes and $F$-statistic:
\begin{gather}
B = \frac{\left(1 + \frac{m-1}{n-m} F\right)^{n/2}}{\sqrt{\frac{m+1}{2n} \prod_{i=1}^m n_i}}
\end{gather}
where $m$ is the number of groups, $n_i$ the sample size within the
$i$th group, $n = \sum n_i$ is the total sample size.  This can be
used in general statistical applications where a Bayesian measure of
evidence is desired, either for retrospective analysis where
$p$-values have been used or where it is convenient to use existing
software like R analysis of variance. We find this frequently useful
as a quick and  easy way to complement a frequentist analysis with a 
Bayesian analysis measure of evidence, that does not seem to be generally realised.

Consider the following R analysis of variance for the \texttt{Oats} dataset:
<<>>=
library(nlme)
data(Oats)
summary(aov(yield ~ nitro* Variety + Error(Block), data=Oats))
with(Oats, table(Variety))
@
Noting that there are 3 groups (3 varieties) of size 24, and an
$F$-statistic of 3.73, the \cite{SpiegelhalterSmith82} Bayes factor
for testing and effect of \texttt{Variety} is calculated as:
<<>>=
SS.oneway.bf(group.sizes=c(24,24,24), Fstat=3.73)
@

A Bayes factor of 2 indicates that the evidence for \texttt{Variety}
is very weak indeed. Traditionally this would have been regarded as
`significant'. Early users of significance tests were fortunate that
their prior odds were generally high.

Note: Although not strictly a one-way analysis of variance, the method
can nevertheless be applied here if it is assumed the $F$-statistic
represents a comparable strength of evidence to the same $F$-statistic
obtained without the additional experimental structure. (This seems
reasonable here, and underlies most frequentist analysis, but don't
take our word for it! We leave it to the reader to find a proof or
derive an adjustment.)

\subsection{\texttt{oneway.bf.alpha()}}
This function, used by \texttt{ld.power()}, calculates the \cite{SpiegelhalterSmith82} 
Bayes factor corresponding to a given $\alpha$ threshold.
For example: the Bayes factors corresponding to $\alpha = 0.05,0.01,0.001$ for testing
for an effect of \texttt{Variety} in the Oats data are calculated as:
<<>>=
oneway.bf.alpha(n=72, group.sizes=c(24,24,24), alpha=c(0.05,0.01,0.001))
@

\subsection{\texttt{calc.alphaB.ABF()}}
This function, used by \texttt{cc.power()} calculates the $\alpha$
threshold corresponding to a given Bayes factor when using asymptotic approximate
Bayes factors, for given prior and sample size. For example, to calculate the
$\alpha$ significance level corresponding to a Bayes factor $B=1000$ for a sample
size $n=20,100,1000$, with prior information equivalent to a single sample point ($a=1$):
<<>>=
calc.alphaB.ABF(B=1000, n=20, a=1, alpha.start=1e-3)
calc.alphaB.ABF(B=1000, n=100, a=1, alpha.start=1e-3)
calc.alphaB.ABF(B=1000, n=1000, a=1, alpha.start=1e-3)
@
Note the decreasing thresholds with increasing sample size. 
Early users of significance tests were fortunate that their prior odds
were generally high and their sample sizes relatively low.

It is sometimes useful to specify an approximate starting value
(\texttt{alpha.start}) as above. The starting value is used as an
initial upper bound in searching for a solution. If the function
returns \texttt{NA}, try a higher or lower starting value. 

The function uses interpolation to find a solution.  The solution can
be refined by choosing a starting value slightly above the solution and using a
smaller value of \texttt{reduction.factor} \eg{}  
<<>>=
calc.alphaB.ABF(B=1000, n=1000, a=1, alpha.start=6e-6,
                reduction.factor=1.05) 
@ 
This makes little difference in this case confirming the accuracy of the
initial solution.

\subsection{\texttt{calc.Balpha.ABF()}}
This function is the inverse of
\texttt{calc.alphaB.ABF()} \ie{}
calculates the approximate Bayes factor corresponding
to a given $\alpha$ level for given prior and sample parameters.
<<>>=
calc.Balpha.ABF(alpha=0.01, n=100, a=1)
calc.alphaB.ABF(B=2.657, n=100, a=1, alpha.start=0.02)
@

Note: Approximate Bayes factors for genetic analysis similar to those
used in \texttt{calc.Balpha.ABF()} were first derived by the author in
\cite[p166--167]{Ball2007AssocMapInPlantsCh8} for S-TDT tests and
subsequently by \cite{WakefieldApproxBF2007} for asymptotically normal
test statistics. Our derivation uses the Savage--Dickey approximation
(\cf{} \cite{Ball2011exptdesigngwcc}), which gives the Bayes factor for
nested models in certain conditions as a ratio of prior to posterior
at zero. Assuming a test statistic with sampling variance $1/n$ and a
prior with mean $0$ and variance $1/a$ equivalent to $a$ sample points
the approximate Bayes factor is given as:
\begin{align}
\label{eq:abf}
B  &\approx \frac{\sqrt{a}}{\sqrt{n+a}}\exp\Bigl(\frac{n^2 Z_n^2}{2(n+a)}\Bigr)
\end{align}

%\section{temp references}
%1. \cite{Altshuleretal2000} \\
%2. \citet{Ball2001} \\
%3. \citep{Ball2005exptdesignbf} \\
%4a. \cite{Ball2007AssocMapInPlantsCh8}\\
%4b.\cite{Ball2007QTLcolocation}\\
%4c. \citep{Ball2007AssocMapInPlantsCh8,Ball2007QTLcolocation}\\
%5. \cite{Ball2009BayesQTLBIC} \\
%6. \cite{Ball2011exptdesigngwcc} \\
%7. \cite{Emahazionetal2001}\\
%8. \cite{HapMap} \\
%9. \cite{OHagan2006prior_elicitation_book} \\
%10. \cite{SillanpaaCorander2002} \\
%10a. \cite{SpiegelhalterSmith82} \\
%11. \cite{TerwilligerWeiss1998} \\
%12. \cite{Weir1996GeneticDataAnalysisII} \\
%13. \cite{YangetalcommonSNPs2010} \\

\bibliographystyle{plainnat}
\bibliography{ldDesign}
\end{document}

\section{References}
\begin{description}
\item[1] Altshuler, D., Hirschhorn, J. N., Klannemark, M., \ldots and
         Lander, E. S. 2000: The common {PPAR$\gamma$} {P}ro12{A}la polymorphism is associated with
                  decreased risk of type 2 diabetes. Nature Genetics \textbf{26}: 76--80.
\item[2] Ball, R.\,D. 2001: Bayesian methods for quantitative
trait loci mapping based on model selection: approximate analysis
using the Bayesian Information Criterion. Genetics \textbf{159:} 1351--1364.

\item[3] Ball, R.\,D. 2005: Experimental designs for reliable detection
  of linkage disequilibrium in unstructured random population
  association studies. Genetics \textbf{170:} 859â€“873.

\item[4] Ball, R.\,D. 2007a: Statistical analysis and experimental
  design, Chapter 8, In: \textit{Association Mapping in Plants}. N.C. Oraguzie
  et al. editors, Springer Verlag, ISBN 0387358447. (69pp)

\item[5] Ball, R.\,D. 2007b: {Q}uantifying evidence for candidate gene
  polymorphisms---{B}ayesian analysis combining sequence-specific and
  {QTL} co-location information. Genetics \textbf{177:} 2399--2416.

\item[6] Ball, R.\,D. 2009: \textit{\texttt{BayesQTLBIC}---Bayesian multi-locus QTL
  analysis based on the BIC criterion.}
  {\small\url{http://cran.r-project.org/web/packages/BayesQTLBIC/index.html}}

\item[7] Ball, R.\,D. 2011: Experimental designs for robust detection of
  effects in genome-wide case-control studies (to appear in Genetics).
  {\small\url{http://www.genetics.org/content/early/2011/09/13/genetics.111.131698.abstract}}

\item[8] Emahazion, T.,  Feuk, L., Jobs, M., \ldots and  Brookes, A.\,J. 2001: SNP 
        association studies in Alzheimer's disease highlight problems for complex disease analysis.
        Trends in Genetics \textbf{17}: 407--413.

\item[9] The HapMap project. \url{www.hapmap.org}.

\item[10] O'Hagan, A. Buck, C.\,E., Daneshkhah, A., Eiser, J.\,R.,
  Garthwaite, P.\,H., Jenkinson, D.\,J., Oakley, J.\,E., and Rakow,
  T. 2006: \textit{Uncertain judgements: eliciting experts'
  probabilities.} Hoboken, NJ: Wiley, xiii+321pp. ISBN:
  978-0-470-02999-2.

\item[11] Sillanp\"a\"a, M.\,J. and Corander, J. 2002: Model choice in
  gene mapping: what and why.  Trends in Genetics {\bf 18}: 301--307.

\item[12] Terwilliger, J.\,D. and Weiss, K.\,M. 1998:
  Linkage disequilibrium mapping of complex disease: fantasy or reality?
  Curr. Opin. Biotechnol. \textbf{9}: 578--594.

\item[13]  Weir, B.\,S. 1996: {\it Genetic Data Analysis II}.
Sinauer Associates, Sunderland MA.

\item[14] Yang, J. Beben, B., McEvoy, D.\,P., Gordon, S.Henders, A.\,K.,
Nyholt, D.\,R., Madden, P.\,A., Heath, A.\,C., Martin, N.\,G.,
Montgomery, G.\,W., Goddard, M.\,E., and Visscher, P.\,M. 2010: Common
SNPs explain a large proportion of the heritability for human height.
Nature Genetics \textbf{42:} 565--569, 608.

\end{description}
\end{document}

TODO: 
rationalise arguments of ld.power, ld.design more similar to cc.power, cc.design
example with 100s of loci 0.1\%--1\%
other functions
oneway.bf.alpha         Correspondence between significance levels and
                        Bayes factors for effects of marker genotype
                        classes.
SS.oneway.bf            Bayes factors for one-way analysis of variance
                        models.
calc.B.ABF              Function to calculate approximate Bayes factor from Z-statistic
                        with sampling distribution N(0,1/n).
calc.Zc.ABF             Function to calculate critical value, Zc for given Bayes factor,
                        prior, and sample parameters.
calc.Zalpha.ABF         Function to calculate Z value corresponding to significance
                        level alpha.
calc.alphaB.ABF         Calculate the alpha value corresponding to a given Bayes factor
                        for given prior and sample parameters.
gpc.power               Function to get power of case control study to detect linkage
                        disequilibrium with a functional locus with a given significance



%\bibliographystyle{plainnat}
%\bibliography{ccDesignVignette}

%Weir book
%Association mapping in plants.
%ldDesign package
%Genetics paper
%O'Hagan book


\end{document}
